### PR TITLE
Worker should exit with ctx.exit to get the right exitcode for non-zero cases

### LINKED ISSUE #7358: Celery worker dies, but has an exit code of 0
# Checklist

- [ ] I have verified that the issue exists against the `master` branch of Celery.
- [ ] This has already been asked to the [discussions forum](https://github.com/celery/celery/discussions) first.
- [ ] I have read the relevant section in the
  [contribution guide](http://docs.celeryproject.org/en/latest/contributing.html#other-bugs)
  on reporting bugs.
- [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [x] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the bug was already fixed in the master branch.
- [x] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information

- [x] I have included the output of ``celery -A proj report`` in the issue.
```

```
- [ ] I have verified that the issue exists against the `master` branch of Celery.
- [ ] I have included the contents of ``pip freeze`` in the issue.
- [ ] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

- [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
- [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
- [ ] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
- [ ] I have tried reproducing the issue on more than one operating system.
- [ ] I have tried reproducing the issue on more than one workers pool.
- [ ] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
- [ ] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues

- None

#### Possible Duplicates

- None

## Environment & Settings

**Celery version**: 5.2.3 (dawn-chorus)

<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```
software -> celery:5.2.3 (dawn-chorus) kombu:5.2.3 py:3.9.10
            billiard:3.6.4.0 py-amqp:5.0.9
platform -> system:Darwin arch:64bit
            kernel version:20.6.0 imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:amqp results:django-db
```

</p>
</details>

# Steps to Reproduce

## Required Dependencies

* **Minimal Python Version**: N/A or Unknown
* **Minimal Celery Version**: N/A or Unknown
* **Minimal Kombu Version**: N/A or Unknown
* **Minimal Broker Version**: N/A or Unknown
* **Minimal Result Backend Version**: N/A or Unknown
* **Minimal OS and/or Kernel Version**: N/A or Unknown
* **Minimal Broker Client Version**: N/A or Unknown
* **Minimal Result Backend Client Version**: N/A or Unknown

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
```

</p>
</details>

### Other Dependencies

<details>
<p>
N/A
</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>

```python
```

</p>
</details>

# Expected Behavior

Expected worker to exit with a non-zero code

# Actual Behavior

Ran celery with the following settings
```
CELERY_BROKER_CONNECTION_RETRY = False
CELERY_BROKER_CONNECTION_TIMEOUT = 2 # seconds
```
because for kubernetes related reasons it's better for us if the workers just die immediately if the config is wrong.

The worker does exit fast enough, but the exit code is 0 despite the following backtrace
```
[2022-03-18 17:33:30,670: CRITICAL/MainProcess] Unrecoverable error: ConnectionRefusedError(61, 'Connection refused')
Traceback (most recent call last):
  File "/Users/tomparker-shemilt/.virtualenvs/case-cards/lib/python3.9/site-packages/celery/worker/worker.py", line 203, in start
    self.blueprint.start(self)
  File "/Users/tomparker-shemilt/.virtualenvs/case-cards/lib/python3.9/site-packages/celery/bootsteps.py", line 116, in start
    step.start(parent)
  File "/Users/tomparker-shemilt/.virtualenvs/case-cards/lib/python3.9/site-packages/celery/bootsteps.py", line 365, in start
    return self.obj.start()
  File "/Users/tomparker-shemilt/.virtualenvs/case-cards/lib/python3.9/site-packages/celery/worker/consumer/consumer.py", line 326, in start
    blueprint.start(self)
  File "/Users/tomparker-shemilt/.virtualenvs/case-cards/lib/python3.9/site-packages/celery/bootsteps.py", line 116, in start
    step.start(parent)
  File "/Users/tomparker-shemilt/.virtualenvs/case-cards/lib/python3.9/site-packages/celery/worker/consumer/connection.py", line 21, in start
    c.connection = c.connect()
  File "/Users/tomparker-shemilt/.virtualenvs/case-cards/lib/python3.9/site-packages/celery/worker/consumer/consumer.py", line 422, in connect
    conn = self.connection_for_read(heartbeat=self.amqheartbeat)
  File "/Users/tomparker-shemilt/.virtualenvs/case-cards/lib/python3.9/site-packages/celery/worker/consumer/consumer.py", line 428, in connection_for_read
    return self.ensure_connected(
  File "/Users/tomparker-shemilt/.virtualenvs/case-cards/lib/python3.9/site-packages/celery/worker/consumer/consumer.py", line 451, in ensure_connected
    conn.connect()
  File "/Users/tomparker-shemilt/.virtualenvs/case-cards/lib/python3.9/site-packages/kombu/connection.py", line 275, in connect
    return self._ensure_connection(
  File "/Users/tomparker-shemilt/.virtualenvs/case-cards/lib/python3.9/site-packages/kombu/connection.py", line 434, in _ensure_connection
    return retry_over_time(
  File "/Users/tomparker-shemilt/.virtualenvs/case-cards/lib/python3.9/site-packages/kombu/utils/functional.py", line 312, in retry_over_time
    return fun(*args, **kwargs)
  File "/Users/tomparker-shemilt/.virtualenvs/case-cards/lib/python3.9/site-packages/kombu/connection.py", line 878, in _connection_factory
    self._connection = self._establish_connection()
  File "/Users/tomparker-shemilt/.virtualenvs/case-cards/lib/python3.9/site-packages/kombu/connection.py", line 813, in _establish_connection
    conn = self.transport.establish_connection()
  File "/Users/tomparker-shemilt/.virtualenvs/case-cards/lib/python3.9/site-packages/kombu/transport/pyamqp.py", line 201, in establish_connection
    conn.connect()
  File "/Users/tomparker-shemilt/.virtualenvs/case-cards/lib/python3.9/site-packages/amqp/connection.py", line 323, in connect
    self.transport.connect()
  File "/Users/tomparker-shemilt/.virtualenvs/case-cards/lib/python3.9/site-packages/amqp/transport.py", line 113, in connect
    self._connect(self.host, self.port, self.connect_timeout)
  File "/Users/tomparker-shemilt/.virtualenvs/case-cards/lib/python3.9/site-packages/amqp/transport.py", line 197, in _connect
    self.sock.connect(sa)
ConnectionRefusedError: [Errno 61] Connection refused
```

I tried messing around with locally editing https://github.com/celery/celery/blob/master/celery/bin/worker.py and noted the
```
        worker.start()
        return worker.exitcode
    except SecurityError as e:
        ctx.obj.error(e.args[0])
        ctx.exit(1)
```

block at the end. If I replaced `return worker.exitcode` with `ctx.exit(worker.exitcode)` it seems to work as I was expecting i.e. exit code of 1

### PR DESCRIPTION
**Pull request by palfrey**
## Description

Fixes #7358

### CODE REVIEW

**Comment by palfrey** [comment]
> good tests!

It's a good test, but doesn't appear to be getting picked up by the test coverage, and despite some work I can't see a way to make a unit test version. I'm not seeing codecov in the CI for the integration tests. I can see two ways to fix this: add the integration tests to code coverage, or merge with a slightly lowered test coverage level. Thoughts?

**Comment by auvipy** [comment]
> > good tests!
> 
> It's a good test, but doesn't appear to be getting picked up by the test coverage, and despite some work I can't see a way to make a unit test version. I'm not seeing codecov in the CI for the integration tests. I can see two ways to fix this: add the integration tests to code coverage, or merge with a slightly lowered test coverage level. Thoughts?

very good points. we do not have extensive integration tests right now. code coverage is small ATM. for unit tests the primary author who refatcor this part of code, left unit test for future which is a over due. as the integration tests are passing we can go with slight lower test coverage.

**Comment by auvipy** [comment]
additonally you can try basic test for worker here https://github.com/celery/celery/tree/master/t/unit/bin using https://pypi.org/project/pytest-click/1.1.0/

**Comment by palfrey** [comment]
> additonally you can try basic test for worker here https://github.com/celery/celery/tree/master/t/unit/bin using https://pypi.org/project/pytest-click/1.1.0/

Thanks, added one using that. Was seeing one integration test failure, but that's with `t/integration/test_canvas.py::test_chord::test_mutable_errback_called_by_chord_from_group_fail_multiple[errback_old_style]` which I'm pretty sure has nothing to do with this. Flaky test?

**Comment by auvipy** [comment]
seems unit tests are failing

File: t/unit/app/test_app.py [hunk 1]
```diff
@@ -591,8 +591,8 @@ def test_worker_main(self, mocked_celery):
         mocked_celery.main.assert_called_with(
             args=['worker', '--help'], standalone_mode=False)
 
-    def test_config_from_envvar(self):
-        os.environ['CELERYTEST_CONFIG_OBJECT'] = 't.unit.app.test_app'
+    def test_config_from_envvar(self, monkeypatch):
+        monkeypatch.setenv("CELERYTEST_CONFIG_OBJECT", 't.unit.app.test_app')
```
Comment (palfrey) [inline t/unit/app/test_app.py:595]: This setting was bleeding through into other tests

**Comment by palfrey** [comment]
> seems unit tests are failing

Yeah. Test was fine on it's own, but the other tests bled through and messed around with the settings, which made it get stuck. I've done some cleanup to reduce that, and it's certainly able to pass with the other earlier tests locally now.

**Comment by auvipy** [comment]
can you recheck why windows builds are failing?

**Comment by palfrey** [comment]
> can you recheck why windows builds are failing?

It looked like it was complaining that my test left a thread running afterwards, but that shouldn't be the case. I haven't got a suitable Windows machine to test with so can't even locally figure out why it only happens there.

**Comment by auvipy** [comment]
t/unit/bin/test_worker.py::test_cli 
  Error: The action has timed out.

should we try increase the timeout duration?

**Comment by palfrey** [comment]
> t/unit/bin/test_worker.py::test_cli Error: The action has timed out.
> 
> should we try increase the timeout duration?

I don't think it's not waiting long enough, I think something is hanging. I've added `--capture=no` to the test args so that we can see the worker config (which might indicate where the problem is) and switched on the solo pool for the test worker to try and remove any dangling thread.

**Review by auvipy — APPROVED** [review]

**Comment by auvipy** [comment]
good job sir!

**Pull request merged**